Add reset preferences page and functionality

ckaiser opened the pull request:
Implements [#4432](https://github.com/aseprite/aseprite/issues/4432), needs some testing in case something fell through the cracks.

dacap commented:
> Unclear if this is a regression or something to do with my install.

It looks like a regression introduced in v1.3.8.

dacap commented:
> Main thing I'm looking for feedback with from @dacap is the changes to the code generation, which I made to avoid having to manually go through every single preference and manually reset everything, and the new virtual methods.

I think it looks good, I would send the implementation/definition of each `optionList()`/`sectionList()` to `pref.xml.cpp` and having only the declaration in `pref.xml.h`.

dacap commented on data/pref.xml:
```diff
@@ -142,7 +142,7 @@
     <section id="general">
       <option id="language" type="std::string" default="&quot;en&quot;" />
-      <option id="screen_scale" type="int" default="0" />
+      <option id="screen_scale" type="int" default="2" />
```
The `screen_scale` is a special case and we're using the default value `0` (probably something to review in the future along with #4606):

https://github.com/aseprite/aseprite/blob/8323a555007e1db9670b098ce4b1b9c5f8b3d7ad/src/app/modules/gui.cpp#L133

We should keep this equal to `0` (and when we reset this value the customization should be removed from the `.ini` and get it from the theme).

ckaiser replied:
I'm now loading the screen and UI scale from the theme, but I agree this should get changed since it's a bit confusing to have a value that sometimes gets clamped or changed like that

dacap commented:
> I also took the liberty of adjusting some of the wording in the preferences page while I was staring at it.

No problem about this, but probably it could be a special commit just for the rewording (sorry for the separated comments/this is the last one for the draft :pray:).

dacap commented:
It looks like 50b284bf4d16c44d4b0eee379e5939d3e93036a8 is reverting the changes from https://github.com/aseprite/aseprite/pull/4629/files

dacap commented:
I'd recommend you to amend/squash your commits and probably rebase to main, I'll review this later when you tell me if it's ready for review.

dacap reviewed (changes requested):
I made some minor comments, didn't have enough time to check all the code and test the feature. But it looks good (separating the signal/connection <-> loadFromPref stuff).

Anyway probably you can check these comments, in general I'd recommend to avoid formatting changes to lines that are not touched when features are implemented as it makes things a little noisy to review. When code lines are moved, probably the best thing to do is to leave them as they were.

ckaiser commented on src/app/recent_files.h:
```diff
@@ -22,13 +23,18 @@ namespace app {
            kCollections };
+
   public:
+    static constexpr std::array<const char*, 4> sections{
+      "PinnedFiles", "RecentFiles", "PinnedPaths", "RecentPaths"
+    };
+
```
I moved it to make it publicly available and easily iterable in a for range, to avoid having the list of sections in multiple places.

dacap replied:
Probably I missed something in this commit, but it looks like the `sections` array is not used outside `src/app/recent_files.cpp`, so we should avoid a change that is not required in the PR. I think we can safely revert this change.

Even if this `sections` constant needs to be public for this PR, we prefer a `k` prefix and `UpperCase` kind of name (like `kSectionName`) for constants (instead of `sections` which can be confused with local variable names).

ckaiser replied:
Oh, my bad, it's a leftover from something I thought I was still using, I've just reverted it.

ckaiser commented on src/app/commands/cmd_options.cpp:
```diff
@@ -625,31 +447,249 @@ class OptionsWindow : public app::gen::Options {
+  void loadFromPreferences()
+  {
+    // Default extension to save files
+    fillExtensionsCombobox(defaultExtension(),
+                           m_pref.saveFile.defaultExtension());
+    fillExtensionsCombobox(exportImageDefaultExtension(),
+                           m_pref.exportFile.imageDefaultExtension());
+    fillExtensionsCombobox(exportAnimationDefaultExtension(),
+                           m_pref.exportFile.animationDefaultExtension());
+    fillExtensionsCombobox(exportSpriteSheetDefaultExtension(),
+                           m_pref.spriteSheet.defaultExtension());
```
Yeah, some stuff triggered on auto-save and I cleaned it up manually but in the wrong direction 😅, I'll do the reverts

dacap replied:
Thanks about this! Probably this is a luxury we can afford now that we don't have the pre-commit configured.

I'm thinking out loud now, but probably in the future, in case some refactor needs to move some code, we can do a full clang-format of the modified files (but this is just for the future).

dacap commented:
I was going to merge this squashed with [some changes](https://gist.github.com/dacap/787b2bdc0c1e9e1d7d498c491cc6b7e0), mainly to keep track/same order of things as the original code (this made it easier to check that everything was still in place).

You can reset your branch to update this PR to this commit: https://github.com/dacap/aseprite/commit/594575188bd11e3e2301c66be45247496af160ff

Anyway it looks that there is still some work to do. I tried just pressing the "Reset" button to reset the configuration (`defaultReset()->isSelected()`) and the language was reset to the first entry in the combobox. You can test this setting the following cmake option: `ENABLE_I18N_STRINGS:BOOL=ON`. This option will clone the strings repo and show you more languages in the combobox of available options (some of these options are included in the official release of Aseprite).

ckaiser commented:
> Anyway it looks that there is still some work to do. I tried just pressing the "Reset" button to reset the configuration (`defaultReset()->isSelected()`) and the language was reset to the first entry in the combobox. You can test this setting the following cmake option: `ENABLE_I18N_STRINGS:BOOL=ON`. This option will clone the strings repo and show you more languages in the combobox of available options (some of these options are included in the official release of Aseprite).

This is intended behavior, should it not be? I assumed that resetting the settings to default also included languages, and the `defaultReset()` option is for the first page where that dropdown lives.

dacap commented:
> This is intended behavior, should it not be? I assumed that resetting the settings to default also included languages, and the `defaultReset()` option is for the first page where that dropdown lives.

Here the issue is that the official version includes some built-in languages that are not extensions, so the Languages combobox will include more items:

![image](https://github.com/user-attachments/assets/1ca2f73e-d3d6-40b8-91fc-ea86672daea2)

In the current state with this PR, just resetting the default setting will bring Afrikaan as the selected language:

![reset-language](https://github.com/user-attachments/assets/032cfd17-ef77-4937-97d7-b4eeeb618dc2)

You can use `ENABLE_I18N_STRINGS:BOOL=ON` cmake option to test this behavior. Remember to continue any work/fix from https://github.com/dacap/aseprite/commit/594575188bd11e3e2301c66be45247496af160ff

ckaiser commented:
Fixed the language issue and rebased this branch's changes on what you had in your fork, I've been a bit sick lately so I haven't had the time to do a full re-test of the PR but I believe it should be good to go now, language stuff skipped me by because I was testing it mostly with language extensions to also test that it was removing them 🙃.

dacap commented:
> I've been a bit sick lately so I haven't had the time to do a full re-test of the PR but I believe it should be good to go now

Take care! I hope you get better soon. I'll test this PR and merge it if it's working :+1:
